File: it.unibo.cautiousExplorer/app/src/main/java/it/unibo/remoteCall/ConnectionReader.java
/*
============================================================
ConnectionReader

============================================================
 */
package it.unibo.remoteCall;

import it.unibo.actor0.ApplMessage;
import it.unibo.interaction.IUniboActor;
import it.unibo.is.interfaces.protocols.IConnInteraction;
import it.unibo.supports2021.ActorBasicJava;
import org.jetbrains.annotations.NotNull;

public class ConnectionReader extends ActorBasicJava  {
    private IConnInteraction conn;

    public ConnectionReader(String name, IConnInteraction conn) {
        super(name);
        this.conn = conn;
    }

    protected void getInput(ConnectionReader obj) throws Exception {
        while (true) {
            //System.out.println("ConnectionReader  | waitInput ... "   );
            String v = conn.receiveALine();
            //System.out.println("ConnectionReader  | getInput " + v );
            if (v != null) {
                String msg = ApplMessage.create(v).toString();
                //System.out.println("ConnectionReader  | updateObservers " + msg );
                updateObservers(msg);
            }else{
                break;
            }
        }//while
    }


    @Override
    protected void handleInput(String s) {
        try {
            //System.out.println("ConnectionReader handleInput:" + s);
            ApplMessage msg = ApplMessage.create(s);
            //System.out.println("ConnectionReader handleInput msg=" + msg.getMsgId());
            if (msg.getMsgId().equals("start")) {
                getInput( this );
            }
        }catch( Exception e){
            System.out.println("ConnectionReader ERROR:" + e.getMessage());
        }

    }//handleInput

    @Override
    public void registerActor(@NotNull IUniboActor iUniboActor) {

    }

    @Override
    public void removeActor(@NotNull IUniboActor iUniboActor) {

    }
}

File: it.unibo.kotlinSupports/app/src/main/kotlin/it/unibo/actor0Remote/ConnectionReader.kt
package it.unibo.actor0Remote

import it.unibo.`is`.interfaces.protocols.IConnInteraction
import it.unibo.actor0.ActorBasicKotlin
import it.unibo.actor0.ApplMessage
import it.unibo.actor0.DispatchType
import it.unibo.actor0.sysUtil
import kotlinx.coroutines.*

class ConnectionReader (name: String, val conn: IConnInteraction ) :
    ActorBasicKotlin( name ) { //DispatchType.single, DispatchType.iobound


    suspend fun getInput( obj : ConnectionReader) {
            while (true) {
                //println("$name  | getInput scope=$scope dispatcher=$dispatcher  ${infoThreads()}")
                val v = conn.receiveALine()
                println("$name | RECEIVES: $v")
                if (v != null) {
                    val msg = ApplMessage.create(v)
                    obj.updateObservers(msg)
                } else{

                    break
                }
            }//while
        println("$name  | TERMINATE -------------------------- ")
            terminate()
    }

    override suspend fun handleInput(msg: ApplMessage) {
        println( "$name  | $msg ${infoThreads()}" )
        if(msg.msgId == "start"){
             getInput( this )  //blocked ...
         }
    }
}